File: SimpleEngine/FileSystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace SimpleEngine
{
    public class FileSystem
    {
        public FileSystem()
        {
        }

        public void addRoot(string root)
        {
            roots.Add(root);
        }

        public void addDefaultRoots(string app, string game)
        {
            addRoot(Environment.CurrentDirectory);
            addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), app, game));
            addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), app, game));
            addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), app, game));
            addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), app, game));
            addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.Programs), app, game));
        }

        private Stream doOpen(string path)
        {
            if (path.Contains("..")) throw new Exception("do not specify parent directory");
            string p = path;//.Replace(':', Path.DirectorySeparatorChar);
            foreach (string r in roots)
            {
                string fp = Path.Combine(r, p);
                if (File.Exists(fp)) return new FileStream(fp, FileMode.Open);
            }
            throw new Exception("Failed to open file " + path);
        }

        public Stream open(string path)
        {
            return doOpen(MapFile(overrides, path));
        }

        private string combine(string s, params string[] args)
        {
            string r = s;
            foreach (string a in args)
            {
                r = Path.Combine(r, a);
            }
            return r;
        }

        List<string> roots = new List<string>();

        public IEnumerable<string> readLines(string path)
        {
            using (Stream s = open(path))
            {
                using (StreamReader sr = new StreamReader(s))
                {
                    while (sr.EndOfStream == false)
                    {
                        yield return sr.ReadLine();
                    }
                }
            }
        }

        Dictionary<string, string> overrides = null;

        internal void setOverrides(Dictionary<string, string> overrides)
        {
            this.overrides = overrides;
        }

        internal void clearOverrides(Dictionary<string, string> overrides)
        {
            this.overrides = null;
        }

        internal static string MapFile(Dictionary<string, string> overrides, string path)
        {
            string finalpath = path.ToLower();
            if (overrides != null)
            {
                while (overrides.ContainsKey(finalpath))
                {
                    finalpath = overrides[finalpath];
                }
            }
            return finalpath;
        }
    }
}

File: src/engine/filesystem.cc
#include "engine/filesystem.h"

#include <cassert>
#include <filesystem>
#include <fstream>
#include <map>
#include <stdexcept>
#include <string>
#include <vector>

#include "engine/fileutil.h"
#include "engine/strings.h"
#include "fmt/core.h"

namespace simple
{
    namespace fs = std::filesystem;

    FileSystem::FileSystem()
    {
    }

    void FileSystem::AddRoot(const std::string& root)
    {
        roots.emplace_back(root);
    }

    void FileSystem::AddDefaultRoots(const std::string& app, const std::string& game)
    {
        AddRoot(fs::current_path().string());
        // addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), app, game));
        // addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), app, game));
        // addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), app, game));
        // addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), app, game));
        // addRoot(combine(Environment.GetFolderPath(Environment.SpecialFolder.Programs), app, game));
    }

    std::string FileSystem::OpenExactPath(const std::string& path) const
    {
        if (path.find("..") != std::string::npos)
        {
            throw std::runtime_error("do not specify parent directory");
        }
        for (auto& r : roots)
        {
            auto fp = fs::path{r} / path;
            if (fs::exists(fp))
            {
                return fp.string();
            }
        }
        throw std::runtime_error(fmt::format("File not found in VFS: '{}'", path));
    }

    std::string FileSystem::Open(const std::string& path) const
    {
        return OpenExactPath(MapFile(overrides, path));
    }

    std::vector<std::string> FileSystem::ReadLines(std::string path) const
    {
        return file_util::LinesIn(Open(path));
    }

    void FileSystem::SetOverrides(const std::map<std::string, std::string>& new_overrides)
    {
        overrides = new_overrides;
    }

    void FileSystem::ClearOverrides(const std::map<std::string, std::string>& new_overrides)
    {
        assert(overrides == new_overrides);
        overrides = {};
    }

    std::string FileSystem::MapFile(const std::map<std::string, std::string>& overrides, const std::string& path)
    {
        std::string finalpath = ToLower(path);

        auto found = overrides.find(finalpath);
        while (found != overrides.end())
        {
            finalpath = found->second;
            found = overrides.find(finalpath);
        }

        return finalpath;
    }
}
